### PR TITLE
Dynamic dropdown with search and custom modules

### PR DESCRIPTION
**Pull request by JamesJeffryes**
Super excited about this! I'm not very familiar with select2 so let me know if I did anything wonky. You can play with the widgets with this narrative: http://ci.kbase.us/narrative/notebooks/ws.33517.obj.1

The search widget is pretty straight forward, the custom one points to a service that only works will full  ids (not a great demonstration) but you can see results with search text like"rxn00001".

### CODE REVIEW

File: kbase-extension/static/kbase/js/widgets/appWidgets2/input/dynamicDropdownInput.js [hunk 1]
```diff
@@ -195,20 +252,32 @@ define([
...
+                        'Size: ' + StringUtil.readableBytes(ret_obj.size) + '<br>',
+                        'Uploaded ' + TimeFormat.getTimeStampStr(ret_obj.mtime, true)
+                    ])
+                ]));
+            } else {
+                var replacer = function (match, p1, offset, string) {return ret_obj[p1]};
+                var formatted_string;
+                if (dd_options.description_template) {
+                    // use slice to avoid modifying global description_template
+                    formatted_string = dd_options.description_template.slice().replace(/{{(.+?)}}/g, replacer)
```
Comment (briehl) [inline kbase-extension/static/kbase/js/widgets/appWidgets2/input/dynamicDropdownInput.js]: missing semicolon.

File: kbase-extension/static/kbase/js/widgets/appWidgets2/input/dynamicDropdownInput.js [hunk 2]
```diff
@@ -195,20 +252,32 @@ define([
...
+                } else {
+                    formatted_string = JSON.stringify(ret_obj)
```
Comment (briehl) [inline kbase-extension/static/kbase/js/widgets/appWidgets2/input/dynamicDropdownInput.js]: also missing semicolon

File: kbase-extension/static/kbase/js/widgets/appWidgets2/input/dynamicDropdownInput.js [hunk 3]
```diff
@@ -227,6 +296,9 @@ define([
                 $(ui.getElement('input-container.input')).select2({
                     templateResult: formatObjectDisplay,
                     templateSelection: function(object) {
+                        if (dd_options.selection_id) {
+                            return object[dd_options.selection_id]
```
Comment (briehl) [inline kbase-extension/static/kbase/js/widgets/appWidgets2/input/dynamicDropdownInput.js]: missing semicolon

File: kbase-extension/static/kbase/js/widgets/appWidgets2/input/dynamicDropdownInput.js [hunk 4]
```diff
@@ -157,9 +197,26 @@ define([
...
+                            results.objects[index] = obj});
+                        return results.objects;
+                    });
             } else {
-                // dynamic service plugin stubs!
-                return [];
+                return Promise.resolve(genericClientCall(call_params))
+                    .then(function (results) {
+                        if (results[0][0]) {
+                            return results[0]
```
Comment (briehl) [inline kbase-extension/static/kbase/js/widgets/appWidgets2/input/dynamicDropdownInput.js]: semicolon

File: kbase-extension/static/kbase/js/widgets/appWidgets2/input/dynamicDropdownInput.js [hunk 5]
```diff
@@ -157,9 +197,26 @@ define([
...
+                        } return []
```
Comment (briehl) [inline kbase-extension/static/kbase/js/widgets/appWidgets2/input/dynamicDropdownInput.js]: the return block should be on a different line. Also, semicolon.

File: kbase-extension/static/kbase/js/widgets/appWidgets2/input/dynamicDropdownInput.js [hunk 6]
```diff
@@ -157,9 +197,26 @@ define([
                         });
                         return results;
                     });
+            } else if (dataSource === 'search') {
+                if (call_params.constructor === Array){
+                    call_params = call_params[0]
```
Comment (briehl) [inline kbase-extension/static/kbase/js/widgets/appWidgets2/input/dynamicDropdownInput.js]: semicolon

File: kbase-extension/static/kbase/js/widgets/appWidgets2/input/dynamicDropdownInput.js [hunk 7]
```diff
@@ -157,9 +197,26 @@ define([
...
+                }
+                return Promise.resolve(searchClient.search_objects(call_params))
+                    .then(function (results) {
+                        results.objects.forEach(function(obj, index) {
+                            obj = flattenObject(obj);
+                            obj.id = obj.guid;
+                            obj.text = obj[dd_options.selection_id];
+                            results.objects[index] = obj});
```
Comment (briehl) [inline kbase-extension/static/kbase/js/widgets/appWidgets2/input/dynamicDropdownInput.js]: It's hard to read where this block ends, the trailing `});`  should be on a separate line.

File: kbase-extension/static/kbase/js/widgets/appWidgets2/input/dynamicDropdownInput.js [hunk 8]
```diff
@@ -141,8 +169,20 @@ define([
             });
         }
 
+        function genericClientCall(call_params) {
+            var swUrl = runtime.config('services.workspace.url').replace('ws', 'service_wizard'),
```
Comment (briehl) [inline kbase-extension/static/kbase/js/widgets/appWidgets2/input/dynamicDropdownInput.js]: There's an actual service_wizard url that should be used. It should be in `services.service_wizard.url`

File: kbase-extension/static/kbase/js/widgets/appWidgets2/input/dynamicDropdownInput.js [hunk 9]
```diff
@@ -50,17 +53,40 @@ define([
...
                 token: runtime.authToken()
             }),
+            searchClient = new KBaseSearchEngine(
+                runtime.config('services.KBaseSearchEngine.url'),
+                {token: runtime.authToken()}
+            ),
             userId = runtime.userId(),
             eventListeners = [];
 
+        var flattenObject = function(ob) {
```
Comment (briehl) [inline kbase-extension/static/kbase/js/widgets/appWidgets2/input/dynamicDropdownInput.js]: What's the function do? Does it just take a nested object and flattens it into key-value pairs? I don't quite follow how it's being used.

File: kbase-extension/static/kbase/js/widgets/appWidgets2/input/dynamicDropdownInput.js [hunk 10]
```diff
@@ -157,9 +197,26 @@ define([
                         });
                         return results;
                     });
+            } else if (dataSource === 'search') {
+                if (call_params.constructor === Array){
```
Comment (briehl) [inline kbase-extension/static/kbase/js/widgets/appWidgets2/input/dynamicDropdownInput.js]: The `Array.isArray()` function is made for this, too.

File: kbase-extension/static/kbase/js/widgets/appWidgets2/input/dynamicDropdownInput.js (hunk 8 above)
Comment (JamesJeffryes) [reply kbase-extension/static/kbase/js/widgets/appWidgets2/input/dynamicDropdownInput.js]: Yeah, bad copy paste.

File: kbase-extension/static/kbase/js/widgets/appWidgets2/input/dynamicDropdownInput.js (hunk 9 above)
Comment (JamesJeffryes) [reply kbase-extension/static/kbase/js/widgets/appWidgets2/input/dynamicDropdownInput.js]: Added a docstring to explain it's function

**Comment by JamesJeffryes** [comment]
PyCharm doesn't suggest semicolons for scopes with a single statement so I often forget them. Still seems to be a good idea for the same reason that you use trailing commas in lists. I'll try to be better about spotting that in future.

**Comment by briehl** [comment]
It's mostly for consistency. Javascript, as a language, doesn't require semicolons. But it's possible to mistakenly make statements on two separate lines that can cause odd behavior. That's not very common, and there's a ton of flame-war blog posts out there that argue one way or another. Mostly, I'm on the side where it's best practice to end a statement with a semicolon, unless there's a compelling reason not to.

File: kbase-extension/static/kbase/js/widgets/appWidgets2/input/dynamicDropdownInput.js [hunk 11]
```diff
@@ -50,17 +53,44 @@ define([
...
             userId = runtime.userId(),
             eventListeners = [];
 
+        /**
+         * This function takes a nested return and returns a flat key-value pairing for use with
+         * handlebar replacement for example {"foo":{"bar": "meh"}} becomes {"foo.bar": "meh"}
+         */
+        var flattenObject = function(ob) {
+            var toReturn = {};
+            for (var i in ob) {
```
Comment (briehl) [inline kbase-extension/static/kbase/js/widgets/appWidgets2/input/dynamicDropdownInput.js]: Is `ob` always expected to be an object? If so, the `Object.keys().forEach()` idiom can be used here, too.

File: kbase-extension/static/kbase/js/widgets/appWidgets2/input/dynamicDropdownInput.js [hunk 12]
```diff
@@ -50,17 +53,44 @@ define([
...
+                if (!ob.hasOwnProperty(i)) continue;
+
+                if ((typeof ob[i]) === 'object') {
```
Comment (briehl) [inline kbase-extension/static/kbase/js/widgets/appWidgets2/input/dynamicDropdownInput.js]: Arrays also return 'object' when used with `typeof`. What should happen with arrays in this function, anyway? Do you expect to see any?

Comment (briehl) [reply kbase-extension/static/kbase/js/widgets/appWidgets2/input/dynamicDropdownInput.js]: I see. That's ok, then.

**Pull request merged**